### DISCUSSION TITLE
単体テストの考え方/使い方 8章

### DISCUSSION THREAD

**Discussion by y-kaneda** [atman-inc/training#103] [General]
輪読会のサマリ用ディスカッション。
8章分の各章単位程度でコメントを記載するので、不明点、相談点、確認点、勉強になったポイントなどは、各自ディスカッションのスレッドへコメントをお願いします。

**Comment by y-kaneda** [comment]
# 第8章 なぜ、統合（integration）テストを行うのか？

**Reply by y-kaneda** [reply]
- ビジネスロジック単体としては正しくても(外部システムと)組み合わせるとおかしいことはよくある

**Comment by y-kaneda** [comment]
## 8.1 統合（integration）テストとは？

**Comment by y-kaneda** [comment]
### 8.1.1 統合テストの役割

**Reply by y-kaneda** [reply]
- 読み進めるにあたりこの本における統合テストの定義と自分の中のイメージを擦り合わせる必要がありそう
- 基本的にコントローラのテストは統合テストだが、全てのプロセス外依存をモックに置き換えれば単体テスト
  - ただし、データベースのプロセス外依存は置き換えるべきではない

**Comment by y-kaneda** [comment]
### 8.1.2 テスト・ピラミッドの振り返り

**Reply by y-kaneda** [reply]
- 単体テストは異常ケースをできるだけ
- 統合テストは基本的な正常ケース(ハッピー・パス)1件と、単体テストで検証できない異常ケース全て

**Reply by nmurakami0** [reply]
- happy path
    - Railsのrequest specでのテストがこれに相当する
    - 400系のレスポンスが返るかチェックするのはedge caseになる？
- プロジェクトの複雑さで、各種類のテストのケース数が変わる
- 統合テスト
    - すべてのプロセス外依存とのやり取りを検証できるような長いhappy pathを見つけ出す
        - 複数のケースに分けてでも、すべてのプロセス外依存とのやり取りを検証すること

**Comment by y-kaneda** [comment]
### 8.1.3 統合テストと早期失敗（Fail Fast）

**Reply by y-kaneda** [reply]
- 統合テストの作成方法
  - 全てのプロセス外依存を通るようなハッピー・パスを見つける
  - 1件で完結できなければ複数
  - 全体として起こるエラー(異常ケース)でも複数の単体テストで確認できる取るに足らないケースの場合は含めない
- 早期失敗
  - DB登録前に登録するデータの準備および検証を全て済ませておいて(何かあったらすぐ中止する)、トランザクションを可能な限り小さくするというようなことが必要かも

**Reply by nmurakami0** [reply]
- Fail Fast
    - 目的
        - フィードバックループの短縮
        - 保存される状態の保護
    - 静的型付け言語で、nullがユースケース的に来ない箇所で、Null許容を止めて例外を発生させるのもFail Fastに入りそう

**Comment by y-kaneda** [comment]
## 8.2 どのようなプロセス外依存をモックに置き換えるべきか？

**Comment by y-kaneda** [comment]
### 8.2.1 2種類のプロセス外依存

**Reply by nmurakami0** [reply]
- 管理下にある依存は実装の詳細にあたる
    - 後方互換を維持する必要はない
    - 統合テストでは、モックに置き換えずにそのまま使うべき
- 管理下にない依存は、観察可能な振る舞いにあたる
    - モックに置き換えれば、後方互換の維持につながる
        - リファクタリングを行っても、コミュニケーション時の使用が壊れていないことを確認出来るため

**Reply by y-kaneda** [reply]
- 管理下にある依存
  - 外部から見られない内部的な依存(実装の詳細)
    - 外部からすれば関心外
  - システム内部のDB等
- 管理下にない依存
  - 外部から見れる(観察可能な振る舞い)
  - テスト対象アプリケーションからはコントロール不可
  - コミュニケーション仕様の維持(後方互換の維持)が必要
  - モック化対象

**Comment by y-kaneda** [comment]
### 8.2.2 管理下にある依存と管理下にない依存の両方の性質を持つプロセス外依存

**Reply by y-kaneda** [reply]
- API使えやと思ったら書いてあった

**Comment by y-kaneda** [comment]
### 8.2.3 統合テストで実際のデータベースを使えない場合

**Reply by y-kaneda** [reply]
- 統合テストでデータベースをモックに置き換えてしまうと、単体テストに加えてリポジトリのメソッドの呼び出し検証ぐらいしかできなくなってしまう
- 統合テストはせずに、ドメイン・モデルの単体テストに専念する

**Comment by y-kaneda** [comment]
## 8.3 どのように統合（integration）テストを行うのか？

**Comment by y-kaneda** [comment]
### 8.3.1 今回テストするシナリオ

**Reply by y-kaneda** [reply]
- 最長のハッピーパス
  - 最も多くの副作用を引き起こすケース
- エッジケースの場合、早期失敗によって副作用が無い場合は統合テストの必要が無い
  - レスポンスでエラーを返すとか見なくていいのか？

**Comment by y-kaneda** [comment]
### 8.3.2 モックに置き換えるべきか否かの依存の分類

**Reply by nmurakami0** [reply]
メッセージ・バスがどうメールを送信するかは管理下にないので、モックに置き換える

**Comment by y-kaneda** [comment]
### 8.3.3 E2E（End-to-End）テストはどうするのか？

**Reply by nmurakami0** [reply]
- E2Eでは、管理下にある依存に直接アクセスして確認しない
  - アプリケーションを介して間接的に確認する
    - e.g., Webアプリならブラウザから実際に確認する

**Reply by y-kaneda** [reply]
- E2Eではクライアントの観点(ユーザーの立場)でテストをし、データベース等の管理下にある依存も直接参照したりはせず、アプリケーションを介して間接的に確認する
  - データベースの値を加工して表示とかあるし

**Comment by y-kaneda** [comment]
### 8.3.4 統合テストの作成

**Comment by y-kaneda** [comment]
## 8.4 インターフェイスを使った依存の抽象化

**Comment by y-kaneda** [comment]
### 8.4.1 インターフェイスと疎結合の関係

**Reply by nmurakami0** [reply]
https://enterprisecraftsmanship.com/posts/ocp-vs-yagni

OCP

> Software entities (classes, modules, functions, etc.) should be open for extension, but closed for modification.

- YAGNI contradicts the Bob Martin’s version of OCP.
- The contradiction is resolved if you put Bob Martin’s OCP to the original Meyer’s perspective. That is if you apply this principle only when your code is used by external teams.
- YAGNI beats Bob Martin’s OCP when you are the only consumer of your code (enterprise software development).
- Bob Martin’s OCP beats YAGNI when you are not the only consumer of your code (3rd party library/framework development).

**Reply by y-kaneda** [reply]
- 実装クラスが1つしかないのであれば抽象化・疎結合やOCP(Open-Closed Principle)の遵守に寄与していない
  - 2つ以上のクラスがあってこそ抽象化が"発見"される
  - YAGNI(You Aren't Gonna Need It)原則から外れる
    - 目に付くと仕様変更等が入り修正が発生する
    - コード量が増えればそれだけコストが発生する

**Comment by y-kaneda** [comment]
### 8.4.2 なぜ、プロセス外依存にインターフェイスを使うのか？

**Reply by nmurakami0** [reply]
- 管理下にない依存に対してのみインターフェースを用意すること
  - モック出来るように
  - たとえそのインターフェースを実装したクラスが1つでも良い

クラスをモックにするのは本書ではアンチパターンで、11章で出てくるとのこと

**Reply by y-kaneda** [reply]
- インターフェースがあるとモックが作れる

**Comment by y-kaneda** [comment]
### 8.4.3 同一プロセス内の依存に対するインターフェイスの使用

**Comment by y-kaneda** [comment]
## 8.5 統合テストのベスト・プラクティス

**Comment by y-kaneda** [comment]
### 8.5.1 ドメイン・モデルの境界を明確にする

**Comment by y-kaneda** [comment]
### 8.5.2 アプリケーションを構成する層を減らす

**Comment by y-kaneda** [comment]
### 8.5.3 循環依存を取り除く

**Comment by y-kaneda** [comment]
### 8.5.4 1つのテスト・ケースに複数の実行（Act）フェーズを用いる場合

**Comment by y-kaneda** [comment]
## 8.6 ログ出力に対するテスト

**Reply by y-kaneda** [reply]
- 概念としての"ログ"の定義を最初にイメージしたほうが良いかも
  - テキストファイルなのか、Cloud Watch Logs 等なのかといった保存方法は関係ない
  - イベント・アクション自体に対しての情報を保存することがログ出力

**Comment by y-kaneda** [comment]
### 8.6.1 そもそも、ログ出力をテストすべきか？

**Reply by y-kaneda** [reply]
- アプリケーションの観察可能な振る舞いの一部であればテストすべき
  - DBに保存したデータと同じで、ユーザーが見るのであれば観察可能な振る舞いの一部

**Reply by nmurakami0** [reply]
- 例えばRedashやMetabaseのようなBIツールでは、ログが管理者権限なら見られたはず
  - そのような場合はサポートログであり、観察可能な振る舞いである

**Comment by y-kaneda** [comment]
### 8.6.2 どのようにログ出力をテストすべきか？

**Reply by y-kaneda** [reply]
- 必要なログ出力のみをテストするため、ログの種類ごとに扱うログのインターフェイスを分ける
  - サポート・ログ(ユーザー向けのイベントログ)
  - 診断ログ(開発者向けのデバッグログ)
- 過度に複雑なコードにならないよう、ログというプロセス外依存とのコミュニケーション部分は分離する

**Reply by nmurakami0** [reply]
- 例えばCLIツールを作る場合はログ出力は結構重要なので、ここに書いているようにLoggerクラスを作ってテストもしやすくすると良いかもしれない
  - ただ、ドメインイベントを使ってまでControllerに寄せるのはあまり実際はやらないかもしれない
- Structured Logというと単にJSONなどで決められた仕様のフォーマット形式になっているログという印象だったが、ここで言っているのはLogstashなどに近いちゃんとしたStructured Logのこと？

**Comment by y-kaneda** [comment]
### 8.6.3 どのくらいのログを出力すれば十分なのか？

**Reply by nmurakami0** [reply]
- これは確かに悩ましい
  - 想定外の問題が起こりうるところだけというのはそうかもしれない

**Comment by y-kaneda** [comment]
### 8.6.4 どのようにログ出力オブジェクト（logger）を受け渡すのか？

**Comment by y-kaneda** [comment]
## 8.7 結論

**Comment by y-kaneda** [comment]
## まとめ
